### PR TITLE
[MRG] Uniformize num_dims and add it for LMNN

### LINKED ISSUE #167: Dimensionality reduction for all algorithms
For all Mahalanobis metric learners, we should be able to reduce the dimension. For those that optimize the transformation matrix `L`, this can be done explicitely by setting the matrix at init to have shape (num_dims, n_features). For the others (that optimize the metric `M`), we could provide the user with `num_dims` which could be set to:
- a number k > n_features: in this case we would do the eigendecomposition of M and would only keep the k components with highest eigenvalues
- similar to scikit-learn's PCA, it could also be a value between 0 and 1 for say some threshold on the eigenvalues, or even a string, for some custom strategy (for instance the elbow rule)

This is the current state in the package: 
- All metric learners that use `transformer_from_metric` (Covariance, LSML, MMC, and SDML) do not have a `num_dims` argument
- All others optimize explicitely `L`, and have a `num_dims` argument (LFDA, MLKR, NCA, RCA) except LMNN, that could have one

Also, should we replace `num_dims` by `n_components`, like this is the case in scikit learn linear transformers ? This is also what we did for this PR on NCA in scikit-learn https://github.com/scikit-learn/scikit-learn/pull/10058

This is also related to #124, since we should check that in the case of a custom matrix for initializing the explicit transformer it is consistent with the desired dimension

### PR DESCRIPTION
**Pull request by wdevazelhes**
This solves #167 partially (only gives a num_dims argument to LMNN and ensures that the check is done uniformly between the algorithms)

### CODE REVIEW

**Comment by wdevazelhes** [comment]
Note: For LMNN, if we have shogun, it will do a pca if asked so, but I couldn't find how the dimension of pca is chosen... Also, maybe we'll not have to bother as soon as we have solved #124, because maybe we will chose that the 'pca' option triggers a scikit-learn PCA in both shogun/not shogun cases...

**Comment by bellet** [comment]
Have you checked what is the pca option of Shogun for? It could be simply a preprocessing applied to the dataset, not an initialization. In which case we will set this option to false when solving #125

**Comment by wdevazelhes** [comment]
> Have you checked what is the pca option of Shogun for? It could be simply a preprocessing applied to the dataset, not an initialization of the algorithm. In which case we will set this option to false when solving #125

Yes, according to this page: http://shogun.ml/notebook/latest/LMNN.html, (in cell 5), they say PCA is used to find the initial matrix

**Comment by bellet** [comment]
> > Have you checked what is the pca option of Shogun for? It could be simply a preprocessing applied to the dataset, not an initialization of the algorithm. In which case we will set this option to false when solving #125
> 
> Yes, according to this page: http://shogun.ml/notebook/latest/LMNN.html, (in cell 5), they say PCA is used to find the initial matrix

Indeed. Looking at this, I do not think they reduce the dimension anyway:
https://shogun-list.shogun-toolbox.narkive.com/n6B3o0Mf/shogun-lmnn-pca-error
But I agree it will be safer to perform the PCA initialization ourselves and pass it to Shogun's train's method

File: metric_learn/_util.py [hunk 1]
```diff
@@ -405,3 +405,15 @@ def validate_vector(u, dtype=None):
   if u.ndim > 1:
     raise ValueError("Input vector should be 1-D.")
   return u
+
+
+def _check_num_dims(n_features, num_dims):
+  """Checks that num_dims is less that n_features and deal with the None
+  case"""
+  if num_dims is None:
+    dim = n_features
```
Comment (perimosocordiae) [inline metric_learn/_util.py]: As a style nitpick, I prefer early-return for cases like this:

```python
if num_dims is None:
  return n_features
if 0 < num_dims <= n_features:
  return num_dims
raise ValueError(...)
```

File: metric_learn/_util.py [hunk 2]
```diff
@@ -405,3 +405,15 @@ def validate_vector(u, dtype=None):
   if u.ndim > 1:
     raise ValueError("Input vector should be 1-D.")
   return u
+
+
+def _check_num_dims(n_features, num_dims):
+  """Checks that num_dims is less that n_features and deal with the None
```
Comment (perimosocordiae) [inline metric_learn/_util.py]: typo: "less than"

File: metric_learn/_util.py [hunk 3]
```diff
@@ -405,3 +405,15 @@ def validate_vector(u, dtype=None):
...
+
+
+def _check_num_dims(n_features, num_dims):
+  """Checks that num_dims is less that n_features and deal with the None
+  case"""
+  if num_dims is None:
+    dim = n_features
+  else:
+    if not 0 < num_dims <= n_features:
+      raise ValueError('Invalid num_dims, must be in [1, %d]' % n_features)
```
Comment (perimosocordiae) [inline metric_learn/_util.py]: Some existing code would only warn if `num_dims > n_features`. I think it's probably better to error out here, but we should keep in mind that this is technically a back-compat break.

File: metric_learn/_util.py (hunk 1 above)
Comment (wdevazelhes) [reply metric_learn/_util.py]: That's seems better I agree, done

File: metric_learn/_util.py (hunk 2 above)
Comment (wdevazelhes) [reply metric_learn/_util.py]: Thanks, done

File: metric_learn/_util.py (hunk 3 above)
Comment (wdevazelhes) [reply metric_learn/_util.py]: That's true, maybe adding it in the changelog is enough ? Something like "for all the algorithms that have a parameter `num_dims`, it will now be checked to be between 1 and `n_features`, with `n_features` the number of dimensions of the input space" ?

Comment (wdevazelhes) [reply metric_learn/_util.py]: I just added it to the release draft

**Comment by wdevazelhes** [comment]
Also, I kept the name `num_dims` for now, but in scikit-learn's `NCA` we went for `n_components`, which respects more scikit-learn's conventions. It would be good to change the name wouldn't it ? If so I'll add deprecation warning for it

**Comment by bellet** [comment]
I agree that `n_components` would be more intuitive, +1 for changing this before the 0.5.0 release with deprecation warning

**Comment by wdevazelhes** [comment]
I investigated the travis fail and there seems to be an issue with some shapes mismatch for LMNN when doing dimensionality reduction. There seems to be a problem with the gradient: Looking into this line the gradient seems to be of the shape of an outer product of Xs (when it should be of the shape of L):

https://github.com/metric-learn/metric-learn/blob/d4badc8adc0a8ce6689dd9a95e89181efaf5ee24/metric_learn/lmnn.py#L196

Looking at the PR on LMNN in scikit-learn, there might be a dot product with L missing:
https://github.com/scikit-learn/scikit-learn/pull/8602/files#diff2914dd42441ed8b594e69b2fe04d23f5R762

Doing this dot product solves the problem
 
I'll look into it

**Comment by wdevazelhes** [comment]
I just pushed a commit that warns the user with a ChangedBehaviorWarning when the init is not set, in the cases when the default init is not the same as it was before, let me know what you think, this happens in the following cases:
- lsml: it used the inverse covariance by default (instead of identity now)
- mmc: it used identity/10 (instead of just identity)
- mlkr: it used pca (instead of 'auto')
- nca: it used np.eye(X.shape[1])/(np.maximum(X.max(axis=0)-X.min(axis=0), EPS) (instead of 'auto')
- sdml: it used the inverse covariance by default (since use_cov was True by default), instead of identity

File: test/test_fit_transform.py [hunk 4]
```diff
@@ -96,9 +96,7 @@ def test_lfda(self):
     res_2 = lfda.fit_transform(self.X, self.y)
 
     # signs may be flipped, that's okay
-    if np.sign(res_1[0,0]) != np.sign(res_2[0,0]):
-        res_2 *= -1
-    assert_array_almost_equal(res_1, res_2)
+    assert_array_almost_equal(abs(res_1), abs(res_2))
```
Comment (wdevazelhes) [inline test/test_fit_transform.py]: The tests were failing because of a sign error, I think it's because of the eigenvalue decomposition in LFDA, that can lead to different signs in the output. I did this modif because I thought that any sign switching of any column still keeps the result valid am I right (any symmetry wrt some of the axis are still valid right?) (Not just a global sign switching as before)? But maybe I should rather try to change the code to make LFDA deterministic ? What do you think @perimosocordiae @bellet ?

Comment (wdevazelhes) [reply test/test_fit_transform.py]: I just raised an issue for that (#211),  I guess maybe let's do like that for now and later fix this non deterministic problem ?

**Comment by wdevazelhes** [comment]
If you agree with my last modifs (see comment https://github.com/metric-learn/metric-learn/pull/193#pullrequestreview-245994079), we should be ready to merge

File: test/test_fit_transform.py (hunk 4 above)
Comment (perimosocordiae) [reply test/test_fit_transform.py]: The sign switch is expected, as each eigenvector may be negated independently of the others. The result is valid in the sense that the underlying optimization problem is still solved, and distances in the transformed space are equivalent. Individual points in the transformed space won't always be the same, though.

I don't think it's worth trying to enforce a deterministic result, but I could probably be persuaded otherwise.

Comment (bellet) [reply test/test_fit_transform.py]: Let's discuss it in the issue you opened
